src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs

```diff
@@ -0,0 +1,553 @@
+
+                    var aggregateId = new JObject();
+
+                    // For the aggregating automat id, we have to provide only the
+                    aggregateId["instanceId"] = automationLogicalId + "/";
+
+                    var aggregateIds = new JArray(aggregateId);
+
+                    run["aggregateIds"] = aggregateIds;
+
+                    modifiedRun = true;
+                }
+            }
+
+            return modifiedRun;
+        }
+
+        private static bool UpdateVersionControlProvenance(JArray runs)
+        {
+            bool modifiedRun = false;
+
+            foreach (JObject run in runs)
+            {
+                JArray versionControlDetailsArray = (JArray)run["versionControlProvenance"];
+                if (versionControlDetailsArray == null) { continue; }
+
+                foreach (JObject versionControlDetails in versionControlDetailsArray)
+                {
+                    modifiedRun |= UpdateVersionControlDetails(versionControlDetails);
+                }
+            }
+            return modifiedRun;
+        }
+
+        private static bool UpdateVersionControlDetails(JObject versionControlDetails)
+        {
+            bool modifiedVersionControlDetails = false;
+
+            // https://github.com/oasis-tcs/sarif-spec/issues/242
+            modifiedVersionControlDetails |= RenameProperty(versionControlDetails, previousName: "timestamp", newName: "asOfTimeUtc");
+
+            // https://github.com/oasis-tcs/sarif-spec/issues/244
+            modifiedVersionControlDetails |= RenameProperty(versionControlDetails, previousName: "uri", newName: "repositoryUri");
+
+            // https://github.com/oasis-tcs/sarif-spec/issues/249
+            modifiedVersionControlDetails |= RenameProperty(versionControlDetails, previousName: "tag", newName: "revisionTag");
+
+            return modifiedVersionControlDetails;
+        }
+
+        internal static bool UpdateRunResults(JArray runs)
+        {
+            bool modifiedRun = false;
+
+            foreach (JObject run in runs)
+            {
+                JArray results = (JArray)run["results"];
+
+                if (results == null) { continue; }
+
+                foreach (JObject result in results)
+                {
+                    // https://github.com/oasis-tcs/sarif-spec/issues/216
+                    //
+                    // result.ruleMessageId is deprecated. This property value
+                    // should be expressed as result.message.messageId instead.
+                    // e.g.:
+                    // "message": { "messageId" : "default"}
+
+                    string ruleMessageId = (string)result["ruleMessageId"];
+
+                    if (ruleMessageId != null)
+                    {
+                        result.Property("ruleMessageId").Remove();
+                        result["message"]["messageId"] = ruleMessageId;
+                        modifiedRun = true;
```

Comment:
>modifiedRun [](start = 24, length = 11)

This is my only quibble with this design: that you have to do this at every change; duplicated code, easy to forget to do it. It would be slower but easier and more reliable to just compare the root-level `JObject`s at the end.

Reply:
Agreed, this is error prone. I need detailed unit tests, will add a todo. The problem with your suggestion is that it is requires a deep clone of each object. we're trying to minimize traversal/object creation.

---
In reply to: [226760895](https://github.com/Microsoft/sarif-sdk/pull/1114#discussion_r226760895) [](ancestors = 226760895)